conans/model/build_info.py

```diff
@@ -553,6 +571,25 @@ def _find_matching(patterns, dirs):
         if self._type != pkg_type:
             out.warning(f"Lib {libname} deduced as '{self._type}, but 'package_type={pkg_type}'")
 
+    def deduce_locations(self, conanfile):
+        if self._exe:   # exe is a new field, it should have the correct location
+            return
+        if self._location or self._link_location:
+            if self._type is None or self._type is PackageType.HEADER:
+                raise ConanException("Incorrect cpp_info defining location without type or header")
+            return
+        if self._type not in [None, PackageType.SHARED, PackageType.STATIC, PackageType.APP]:
+            return
+
+        if len(self.libs) == 0:
+            return
+
+        if len(self.libs) != 1:
+            raise ConanException("More than 1 library defined in cpp_info.libs, cannot deduce CPS")
+
+        # If no location is defined, it's time to guess the location
+        self._auto_deduce_locations(conanfile)
```

Comment:
I think this code would be a bit more clear like this?

```suggestion
    num_libs = len(self.libs)
    if num_libs == 0:
        return
    elif num_libs > 1:
        raise ConanException(f"More than 1 library defined in cpp_info.libs, cannot deduce CPS ({num_libs} libraries found)")
    else:
        self._auto_deduce_locations(conanfile)
```